Classes/Private/Manager/Video/OCTVideoEngine.m

```diff
@@ -125,22 +125,24 @@ - (BOOL)isSendingVideo
     return self.captureSession.isRunning;
 }
 
-- (CALayer *)videoCallPreview
+- (void)getVideoCallPreview:(void (^)(CALayer *))completionBlock
 {
     DDLogVerbose(@"%@: videoCallPreview", self);
-    AVCaptureVideoPreviewLayer *previewLayer = [AVCaptureVideoPreviewLayer layerWithSession:self.captureSession];
-
-    return previewLayer;
+    dispatch_async(self.processingQueue, ^{
+        AVCaptureVideoPreviewLayer *previewLayer = [AVCaptureVideoPreviewLayer layerWithSession:self.captureSession];
+        completionBlock(previewLayer);
+    });
 }
 
-- (UIView *)videoFeedWithRect:(CGRect)rect;
+- (void)getVideoFeed:(void (^)(UIView *videoFeed))completionBlock
 {
-    DDLogVerbose(@"%@: videoFeedWithRect", self);
-    if (! self.videoView) {
-        self.videoView = [[OCTVideoView alloc] initWithFrame:rect];
-    }
-
-    return self.videoView;
+    DDLogVerbose(@"%@: getVideoFeed", self);
+    dispatch_async(self.processingQueue, ^{
+        if (! self.videoView) {
+            self.videoView = [[OCTVideoView alloc] initWithFrame:CGRectZero];
+        }
+        completionBlock(self.videoView);
```

Comment:
I think it would be better to return feed and preview on main queue. So you can do dispatch_async once again. This way there will be no need for changing queue in client's code.

Also I'm wondering, wouldn't it be better to pefrorm sync operation and return preview/feed? Or it may block?

```
- (UIView *)videoFeed
{
    __block UIView *feed;

    dispatch_sync(...{
        feed = ...
    }
    return feed;
}
```